session.go

```diff
@@ -1584,6 +1616,31 @@ func (t *traceWriter) Trace(traceId []byte) {
 	}
 }
 
+type QueryObservation struct {
+	ctx      context.Context // the query's context. Use it to pass arbitrary data to the collector
+	keyspace string
+	stmt     string
+
+	start time.Time // time immediately before the query was called
+	end   time.Time // time immediately after the query returned
+
+	rows int // the number of rows in the current iter. In multi-scans, rows from previous scans are not counted
+
+	err error
+}
+
+// TODO - BatchObservation, similar to QueryObservation but adapted to Batch
+
+// QueryObserver is the interface implemented by query observers / stat collectors.
+type QueryObserver interface {
+	// Observe gets called on every query to cassandra, including all queries in an iterator when paging is enabled.
+	// It doesn't get called if there is no query because the session is closed or there are no connections available.
+	// The error reported only shows query errors, i.e. if a SELECT is valid but finds no matches it will be nil.
+	Observe(QueryObservation)
```

Comment:
Call this ObserveQuery instead and change the signature to take a context as the first param